drafts/inlinemacros.md

```diff
@@ -80,39 +86,37 @@ by-value parameters and that are not side-effect free paths. Arguments to by-nam
 parameters are left alone.
 
    These transformation are intended to preserve the semantics
-of function applications under inlining: An  function call should have the same
-semantics independently on whether the function was maked `inline` or not.
+of function applications under inlining: A function call should have the same
+semantics with respect to side effects independently on whether the function was made `inline` or not.
+
+   After the rewriting is complete, the result of the rewriting is forcibly upcast
+to the type of the original expression. This mechanism works exactly the same 
+as the mechanism of blackbox macro expansion in Scala 2.10+.
 
    The rewriting is done on typed trees. Any references from the function
 body to its environment will be kept in the rewritten code. Accessors
 will be added as needed to ensure that private definitions can be
 accessed from the callsite.
 
-2. A conditional expression `if (c) t else e` where `c` has type `inline Boolean`
+   The rewriting is done in the "outside in" style, i.e. calls to inline methods
+are expanded before possible calls to inline methods in their prefixes and arguments.
+This is different from how the "inside out" style of macro expansion in Scala 2.10+, 
+where prefixes and arguments are expanded first. The old style of macro expansion
+can, if necessary, be emulated by the new style of inline rewritings.
+
+2. If `prefix.f[Ts](args1)...(argsN)` refers to a partially applied inline
+method, an error is raised. Eta expansion of inline methods is prohibited.
+
+3. A conditional expression `if (c) t else e` where `c` has type `inline Boolean`
 is replaced by `t` if `c == true` and by `e` otherwise.
 
-   Note: We might also decide reduce `match` expressions with inline
+   Note: We might also decide to reduce `match` expressions with inline
 selectors and inline patterns. While useful, this would be much harder
 to spec and implement than conditionals.
 
-3. A field selection of an inline class with an inline argument selects the argument: E.g.,
+4. A field selection of an inline class with an inline argument selects the argument: E.g.,
 `new C(42).x` rewrites to `42`.
 
-Inline expressions are rewritten "outside in".
-
-
-## Inline Default Arguments
-
-A default argument for an effectively final method may be marked `inline`. In that case,
-the argument expression is supplied directly at the call site, If the method does not
-override any other method, the usual generation of a "default method" can be omitted.
-
-Example: Given
-
-     final def foo(x: Int)(y: Int = inline x + 1) = ...
-
-Then `foo(0)()` expands to `foo(0)(0 + 1)`
-
```

Comment:
This is not a suggestion to remove inline default arguments, but a call for clarification. There is a number of unclear aspects of this section, mostly in how it avoid talking about side effects, and we might want to address that.

```
def foo(x: Int)(y: Int = x.+(1)): Nothing = scala.this.Predef.???;

// compiles to
def foo(x: Int)(y: Int): Nothing = ...
<synthetic> def foo$default$2(x: Int): Int = x.+(1);

```
foo(0)()

// expands to:
{
  <artifact> val x$1: Int = 0;
  <artifact> val x$2: Int = Test.this.foo$default$2(x$1);
  Test.this.foo(x$1)(x$2)
}
```